cpu/efm32/periph/i2c.c

```diff
@@ -122,128 +126,102 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
 
 int i2c_acquire(i2c_t dev)
 {
+    /* acquire lock */
     mutex_lock(&i2c_lock[dev]);
 
+    /* power peripheral */
+    CMU_ClockEnable(i2c_config[dev].cmu, true);
+
     return 0;
 }
 
 int i2c_release(i2c_t dev)
 {
+    /* disable peripheral */
+    CMU_ClockEnable(i2c_config[dev].cmu, false);
+
+    /* release lock */
     mutex_unlock(&i2c_lock[dev]);
 
     return 0;
 }
 
-int i2c_read_byte(i2c_t dev, uint8_t address, void *data)
+int i2c_read_byte(i2c_t dev, uint16_t address, void *data, uint8_t flags)
 {
-    return i2c_read_bytes(dev, address, data, 1);
+    return i2c_read_bytes(dev, address, data, 1, flags);
 }
 
-int i2c_read_bytes(i2c_t dev, uint8_t address, void *data, int length)
+int i2c_read_bytes(i2c_t dev, uint16_t address, void *data, size_t length, uint8_t flags)
 {
     I2C_TransferSeq_TypeDef transfer;
 
     transfer.addr = (address << 1);
-    transfer.flags = I2C_FLAG_READ;
+    transfer.flags = I2C_FLAG_READ | (flags & I2C_ADDR10 ? I2C_FLAG_10BIT_ADDR : 0);
     transfer.buf[0].data = (uint8_t *) data;
     transfer.buf[0].len = length;
 
     /* start a transfer */
-    _transfer(dev, &transfer);
-
-    if (i2c_progress[dev] != i2cTransferDone) {
-        return -2;
-    }
-
-    return length;
+    return _transfer(dev, &transfer);
 }
 
-int i2c_read_reg(i2c_t dev, uint8_t address, uint8_t reg, void *data)
+int i2c_read_reg(i2c_t dev, uint16_t address, uint16_t reg, void *data, uint8_t flags)
```

Comment:
`i2c_read_reg` and `i2c_write_reg` are now provided by the common driver in `drivers/periph_common`. Just define `PERIPH_I2C_NEED_READ_REG` and `PERIPH_I2C_NEED_WRITE_REG` in the `periph_cpu.h` file of your CPU type.
For example in #9202, I added the following in `cpu/stm32_common/periph_cpu_common.h`:

```c
/**
 * @name    Use the shared I2C functions
 * @{
 */
/** Use read regs function from periph common */
#define PERIPH_I2C_NEED_READ_REG
/** Use write regs function from periph common */
#define PERIPH_I2C_NEED_WRITE_REG
/** @} */
```

Other defines exist for `PERIPH_I2C_NEED_READ_REGS` and `PERIPH_I2C_NEED_WRITE_REGS` if your CPU driver implementation works with them.

Reply:
@aabadie I don't want to use the common implementations, because I then have to use the optional I2C flags, which I cannot use with `emlib`. Emlib has its own way to provide `I2C_START` and `I2C_STOP`.

Reply:
I see. Why not converting RIOT common flags to emlib one (using a switch) ? This way you can get rid of a these functions.

Reply:
I can see if that works, but it would require me to briefly patch [`em_i2c.c`](https://github.com/basilfx/RIOT-gecko-sdk/blob/master/dist/emlib/src/em_i2c.c) (which I don't mind, since Gecko-SDK is already prepared for RIOT-OS).